### PR TITLE
Refine docs in `BaseStorage`

### PR DESCRIPTION
**Pull request by wattlebirdaz**
## Motivation
Documents in `BaseStorage` needed some refinements because they seemed to be a bit misleading. The document described the characteristics of  `CachedStorage` for most of the part and it was not clear what requirement storage classes implementing `BaseStorage` must hold.

## Description of the changes
- Moved some paragraphs from `BaseStorage` to `CachedStorage`.

### CODE REVIEW

**Comment by c-bata** [comment]
Let me reassign reviewer since Imamura-san will take a vacation.

@knshnb Could you review this PR?

File: optuna/storages/_base.py [hunk 1]
```diff
@@ -24,64 +24,27 @@ class BaseStorage(object, metaclass=abc.ABCMeta):
...
 
+    A storage class implementing this class must meet the following requirements.
+
     **Thread safety**
 
-    A storage class can be shared among multiple threads, and must therefore be thread-safe.
-    It must guarantee that return values such as `FrozenTrial`s are never modified.
-    A storage class can assume that return values are never modified by its user.
+    A storage class instance can be shared among multiple threads, and must therefore be
+    thread-safe. It must guarantee that data instance read from the storage must not be modified
```
Comment (keisuke-umezawa) [inline optuna/storages/_base.py:35]: `a data instance` or `data instances`

File: optuna/storages/_base.py [hunk 2]
```diff
@@ -24,64 +24,27 @@ class BaseStorage(object, metaclass=abc.ABCMeta):
...
+    by subsequent writes. For example, `FrozenTrial` instance returned by `get_trial`
+    should not be updated by the subsequent `set_trial_xxx`. This is usually achieved by replacing
+    the old data with a copy on `set_trial_xxx`.
+
+    A storage class can also assume that data instance returned are never modified by its user.
```
Comment (keisuke-umezawa) [inline optuna/storages/_base.py:40]: a data instance or data instances

File: optuna/storages/_cached_storage.py [hunk 3]
```diff
@@ -42,18 +42,41 @@ class _CachedStorage(BaseStorage, BaseHeartbeat):
...
+    are guaranteed to return the same or more recent values than the value at the time of the
+    call to the method.
+
+    Let `T` be a `Trial`. Let `P` be the process that last updated the `state` attribute of `T`.
     Then, any reads on any attributes of `T` are guaranteed to return the same or
     more recent values than any writes by `P` on the attribute before `P` updated
     the `state` attribute of `T`.
     The same applies for `user_attrs', 'system_attrs' and 'intermediate_values` attributes.
 
+    Current implementation of :class:`~optuna.storages._CachedStorage` assumes that each
```
Comment (keisuke-umezawa) [inline optuna/storages/_cached_storage.py:63]:
```suggestion
    The current implementation of :class:`~optuna.storages._CachedStorage` assumes that each
```

**Review by keisuke-umezawa — APPROVED** [review]
I left some comments from grammarly. The other lines look good to me. Please merge it after updating it.

**Comment by wattlebirdaz** [comment]
Thank you for the comments @keisuke-umezawa!
I updated the doc-string.

**Review by c-bata — APPROVED** [review]
LGTM! Thank you for your contribution.

**Pull request merged**